src/psyclone/dynamo0p3.py

```diff
@@ -1472,14 +1510,447 @@ def declare_cma_ops(self, parent):
+                        self._diff_basis_fns.append(entry)
+
+    def initialise_basis_fns(self, parent):
+        ''' Create the declarations and assignments required for the
+        basis-functions required by an invoke '''
+        from psyclone.f2pygen import CommentGen, AssignGen, DeclGen, \
+            AllocateGen, UseGen
+        var_dim_list = []
+        basis_declarations = []
+        op_name_list = []
+
+        if self._qr_vars:
+            parent.add(CommentGen(parent, ""))
+            parent.add(CommentGen(parent, " Look-up quadrature variables"))
+            parent.add(CommentGen(parent, ""))
+
+        for qr_var_name in self._qr_vars:
+            # We generate unique names for the integers holding the numbers
+            # of quadrature points by appending the name of the quadrature
+            # argument
+            qr_vars = ["nqp_h", "nqp_v"]
+            qr_ptr_vars_1d = {"zp": "xqp_v", "wh": "wqp_h",
+                              "wv": "wqp_v"}
+            # Python 2 has no one-line way of combining dicts
+            qr_ptr_vars = qr_ptr_vars_1d.copy()
+            qr_ptr_vars.update({"xp": "xqp_h"})
+            parent.add(
+                DeclGen(
+                    parent, datatype="integer",
+                    entity_decls=[name+"_"+qr_var_name for name in qr_vars]))
+            parent.add(
+                DeclGen(
+                    parent, datatype="real", pointer=True, kind="r_def",
+                    entity_decls=["xp"+"_"+qr_var_name+"(:,:) => null()"]))
+            decl_list = [name+"_"+qr_var_name+"(:) => null()"
+                         for name in qr_ptr_vars_1d]
+            parent.add(
+                DeclGen(parent, datatype="real", pointer=True,
+                        kind="r_def", entity_decls=decl_list))
+            for qr_var in qr_ptr_vars:
+                parent.add(
+                    AssignGen(parent, pointer=True, lhs=qr_var+"_"+qr_var_name,
+                              rhs=qr_var_name + "%get_" +
+                              qr_ptr_vars[qr_var] + "()"))
+            for qr_var in qr_vars:
+                parent.add(
+                    AssignGen(parent, lhs=qr_var+"_"+qr_var_name,
+                              rhs=qr_var_name + "%get_" + qr_var + "()"))
+
+        if self._unique_evaluator_args:
+            parent.add(UseGen(parent, name="function_space_mod",
+                              only=True, funcnames=["BASIS", "DIFF_BASIS"]))
+            parent.add(CommentGen(parent, ""))
+            parent.add(CommentGen(parent,
+                                  " Initialise evaluator-related quantities "
+                                  "using the field(s) that are written to"))
+            parent.add(CommentGen(parent, ""))
+
+        for arg in self._unique_evaluator_args:
+            # We need an 'ndf_nodal' for each unique FS for which there
+            # is an evaluator
+
+            if arg.is_operator:
```

Comment:
Good idea. I've added `DynKernelArgument.evaluator_function_space` that does this. (I thought about doing an internal function but then it would have had no use of `self` which didn't seem right.)